// @aws-cdk/pipelines/test/build-role-policy-statements.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
require("@aws-cdk/assert/jest");
const codepipeline = require("@aws-cdk/aws-codepipeline");
const aws_iam_1 = require("@aws-cdk/aws-iam");
const core_1 = require("@aws-cdk/core");
const cdkp = require("../lib");
const testutil_1 = require("./testutil");
let app;
let pipelineStack;
let sourceArtifact;
let cloudAssemblyArtifact;
beforeEach(() => {
    app = new testutil_1.TestApp();
    pipelineStack = new core_1.Stack(app, 'PipelineStackPolicy', { env: testutil_1.PIPELINE_ENV });
    sourceArtifact = new codepipeline.Artifact();
    cloudAssemblyArtifact = new codepipeline.Artifact('CloudAsm');
});
afterEach(() => {
    app.cleanup();
});
test('Build project includes codeartifact policy statements for role', () => {
    // WHEN
    new testutil_1.TestGitHubNpmPipeline(pipelineStack, 'Cdk', {
        sourceArtifact,
        cloudAssemblyArtifact,
        synthAction: cdkp.SimpleSynthAction.standardNpmSynth({
            sourceArtifact,
            cloudAssemblyArtifact,
            rolePolicyStatements: [
                new aws_iam_1.PolicyStatement({
                    actions: ['codeartifact:*', 'sts:GetServiceBearerToken'],
                    resources: ['arn:my:arn'],
                }),
            ],
        }),
    });
    // THEN
    expect(pipelineStack).toHaveResourceLike('AWS::IAM::Policy', {
        PolicyDocument: {
            Statement: assert_1.arrayWith(assert_1.deepObjectLike({
                Action: [
                    'codeartifact:*',
                    'sts:GetServiceBearerToken',
                ],
                Resource: 'arn:my:arn',
            })),
        },
    });
});
