// @aws-cdk/pipelines/test/cross-environment-infra.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
require("@aws-cdk/assert/jest");
const core_1 = require("@aws-cdk/core");
const testutil_1 = require("./testutil");
let app;
let pipelineStack;
let pipeline;
beforeEach(() => {
    app = new testutil_1.TestApp();
    pipelineStack = new core_1.Stack(app, 'PipelineStack', { env: testutil_1.PIPELINE_ENV });
    pipeline = new testutil_1.TestGitHubNpmPipeline(pipelineStack, 'Cdk');
});
afterEach(() => {
    app.cleanup();
});
test('in a cross-account/cross-region setup, artifact bucket can be read by deploy role', () => {
    // WHEN
    pipeline.addApplicationStage(new TestApplication(app, 'MyApp', {
        env: { account: '321elsewhere', region: 'us-elsewhere' },
    }));
    // THEN
    app.synth();
    const supportStack = app.node.findAll().filter(core_1.Stack.isStack).find(s => s.stackName === 'PipelineStack-support-us-elsewhere');
    expect(supportStack).not.toBeUndefined();
    expect(supportStack).toHaveResourceLike('AWS::S3::BucketPolicy', {
        PolicyDocument: {
            Statement: assert_1.arrayWith(assert_1.objectLike({
                Action: assert_1.arrayWith('s3:GetObject*', 's3:GetBucket*', 's3:List*'),
                Principal: {
                    AWS: {
                        'Fn::Join': ['', [
                                'arn:',
                                { Ref: 'AWS::Partition' },
                                assert_1.stringLike('*-deploy-role-*'),
                            ]],
                    },
                },
            })),
        },
    });
});
test('in a cross-account/same-region setup, artifact bucket can be read by deploy role', () => {
    // WHEN
    pipeline.addApplicationStage(new TestApplication(app, 'MyApp', {
        env: { account: '321elsewhere', region: testutil_1.PIPELINE_ENV.region },
    }));
    // THEN
    expect(pipelineStack).toHaveResourceLike('AWS::S3::BucketPolicy', {
        PolicyDocument: {
            Statement: assert_1.arrayWith(assert_1.objectLike({
                Action: ['s3:GetObject*', 's3:GetBucket*', 's3:List*'],
                Principal: {
                    AWS: {
                        'Fn::Join': ['', [
                                'arn:',
                                { Ref: 'AWS::Partition' },
                                assert_1.stringLike('*-deploy-role-*'),
                            ]],
                    },
                },
            })),
        },
    });
});
/**
 * Our application
 */
class TestApplication extends core_1.Stage {
    constructor(scope, id, props) {
        super(scope, id, props);
        new testutil_1.BucketStack(this, 'Stack');
    }
}
