// @aws-cdk/pipelines/test/pipeline.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
require("@aws-cdk/assert/jest");
const cp = require("@aws-cdk/aws-codepipeline");
const cpa = require("@aws-cdk/aws-codepipeline-actions");
const core_1 = require("@aws-cdk/core");
const testutil_1 = require("./testutil");
let app;
let pipelineStack;
let pipeline;
beforeEach(() => {
    app = new testutil_1.TestApp();
    pipelineStack = new core_1.Stack(app, 'PipelineStack', { env: testutil_1.PIPELINE_ENV });
    pipeline = new testutil_1.TestGitHubNpmPipeline(pipelineStack, 'Cdk');
});
afterEach(() => {
    app.cleanup();
});
test('references stack template in subassembly', () => {
    // WHEN
    pipeline.addApplicationStage(new OneStackApp(app, 'App'));
    // THEN
    expect(pipelineStack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
        Stages: assert_1.arrayWith({
            Name: 'App',
            Actions: assert_1.arrayWith(assert_1.objectLike({
                Name: 'Stack.Prepare',
                InputArtifacts: [assert_1.objectLike({})],
                Configuration: assert_1.objectLike({
                    StackName: 'App-Stack',
                    TemplatePath: assert_1.stringLike('*::assembly-App/*.template.json'),
                }),
            })),
        }),
    });
});
test('action has right settings for same-env deployment', () => {
    // WHEN
    pipeline.addApplicationStage(new OneStackApp(app, 'Same'));
    // THEN
    expect(pipelineStack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
        Stages: assert_1.arrayWith({
            Name: 'Same',
            Actions: [
                assert_1.objectLike({
                    Name: 'Stack.Prepare',
                    RoleArn: {
                        'Fn::Join': ['', [
                                'arn:',
                                { Ref: 'AWS::Partition' },
                                ':iam::',
                                { Ref: 'AWS::AccountId' },
                                ':role/cdk-hnb659fds-deploy-role-',
                                { Ref: 'AWS::AccountId' },
                                '-',
                                { Ref: 'AWS::Region' },
                            ]],
                    },
                    Configuration: assert_1.objectLike({
                        StackName: 'Same-Stack',
                        RoleArn: {
                            'Fn::Join': ['', [
                                    'arn:',
                                    { Ref: 'AWS::Partition' },
                                    ':iam::',
                                    { Ref: 'AWS::AccountId' },
                                    ':role/cdk-hnb659fds-cfn-exec-role-',
                                    { Ref: 'AWS::AccountId' },
                                    '-',
                                    { Ref: 'AWS::Region' },
                                ]],
                        },
                    }),
                }),
                assert_1.objectLike({
                    Name: 'Stack.Deploy',
                    RoleArn: {
                        'Fn::Join': ['', [
                                'arn:',
                                { Ref: 'AWS::Partition' },
                                ':iam::',
                                { Ref: 'AWS::AccountId' },
                                ':role/cdk-hnb659fds-deploy-role-',
                                { Ref: 'AWS::AccountId' },
                                '-',
                                { Ref: 'AWS::Region' },
                            ]],
                    },
                    Configuration: assert_1.objectLike({
                        StackName: 'Same-Stack',
                    }),
                }),
            ],
        }),
    });
});
test('action has right settings for cross-account deployment', () => {
    // WHEN
    pipeline.addApplicationStage(new OneStackApp(app, 'CrossAccount', { env: { account: 'you' } }));
    // THEN
    expect(pipelineStack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
        Stages: assert_1.arrayWith({
            Name: 'CrossAccount',
            Actions: [
                assert_1.objectLike({
                    Name: 'Stack.Prepare',
                    RoleArn: {
                        'Fn::Join': ['', [
                                'arn:',
                                { Ref: 'AWS::Partition' },
                                ':iam::you:role/cdk-hnb659fds-deploy-role-you-',
                                { Ref: 'AWS::Region' },
                            ]],
                    },
                    Configuration: assert_1.objectLike({
                        StackName: 'CrossAccount-Stack',
                        RoleArn: {
                            'Fn::Join': ['', [
                                    'arn:',
                                    { Ref: 'AWS::Partition' },
                                    ':iam::you:role/cdk-hnb659fds-cfn-exec-role-you-',
                                    { Ref: 'AWS::Region' },
                                ]],
                        },
                    }),
                }),
                assert_1.objectLike({
                    Name: 'Stack.Deploy',
                    RoleArn: {
                        'Fn::Join': ['', [
                                'arn:',
                                { Ref: 'AWS::Partition' },
                                ':iam::you:role/cdk-hnb659fds-deploy-role-you-',
                                { Ref: 'AWS::Region' },
                            ]],
                    },
                    Configuration: assert_1.objectLike({
                        StackName: 'CrossAccount-Stack',
                    }),
                }),
            ],
        }),
    });
});
test('action has right settings for cross-region deployment', () => {
    // WHEN
    pipeline.addApplicationStage(new OneStackApp(app, 'CrossRegion', { env: { region: 'elsewhere' } }));
    // THEN
    expect(pipelineStack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
        Stages: assert_1.arrayWith({
            Name: 'CrossRegion',
            Actions: [
                assert_1.objectLike({
                    Name: 'Stack.Prepare',
                    RoleArn: {
                        'Fn::Join': ['', [
                                'arn:',
                                { Ref: 'AWS::Partition' },
                                ':iam::',
                                { Ref: 'AWS::AccountId' },
                                ':role/cdk-hnb659fds-deploy-role-',
                                { Ref: 'AWS::AccountId' },
                                '-elsewhere',
                            ]],
                    },
                    Region: 'elsewhere',
                    Configuration: assert_1.objectLike({
                        StackName: 'CrossRegion-Stack',
                        RoleArn: {
                            'Fn::Join': ['', [
                                    'arn:',
                                    { Ref: 'AWS::Partition' },
                                    ':iam::',
                                    { Ref: 'AWS::AccountId' },
                                    ':role/cdk-hnb659fds-cfn-exec-role-',
                                    { Ref: 'AWS::AccountId' },
                                    '-elsewhere',
                                ]],
                        },
                    }),
                }),
                assert_1.objectLike({
                    Name: 'Stack.Deploy',
                    RoleArn: {
                        'Fn::Join': ['', [
                                'arn:',
                                { Ref: 'AWS::Partition' },
                                ':iam::',
                                { Ref: 'AWS::AccountId' },
                                ':role/cdk-hnb659fds-deploy-role-',
                                { Ref: 'AWS::AccountId' },
                                '-elsewhere',
                            ]],
                    },
                    Region: 'elsewhere',
                    Configuration: assert_1.objectLike({
                        StackName: 'CrossRegion-Stack',
                    }),
                }),
            ],
        }),
    });
});
test('action has right settings for cross-account/cross-region deployment', () => {
    // WHEN
    pipeline.addApplicationStage(new OneStackApp(app, 'CrossBoth', { env: { account: 'you', region: 'elsewhere' } }));
    // THEN
    expect(pipelineStack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
        Stages: assert_1.arrayWith({
            Name: 'CrossBoth',
            Actions: [
                assert_1.objectLike({
                    Name: 'Stack.Prepare',
                    RoleArn: {
                        'Fn::Join': ['', [
                                'arn:',
                                { Ref: 'AWS::Partition' },
                                ':iam::you:role/cdk-hnb659fds-deploy-role-you-elsewhere',
                            ]],
                    },
                    Region: 'elsewhere',
                    Configuration: assert_1.objectLike({
                        StackName: 'CrossBoth-Stack',
                        RoleArn: {
                            'Fn::Join': ['', [
                                    'arn:',
                                    { Ref: 'AWS::Partition' },
                                    ':iam::you:role/cdk-hnb659fds-cfn-exec-role-you-elsewhere',
                                ]],
                        },
                    }),
                }),
                assert_1.objectLike({
                    Name: 'Stack.Deploy',
                    RoleArn: {
                        'Fn::Join': ['', [
                                'arn:',
                                { Ref: 'AWS::Partition' },
                                ':iam::you:role/cdk-hnb659fds-deploy-role-you-elsewhere',
                            ]],
                    },
                    Region: 'elsewhere',
                    Configuration: assert_1.objectLike({
                        StackName: 'CrossBoth-Stack',
                    }),
                }),
            ],
        }),
    });
});
test('pipeline has self-mutation stage', () => {
    // THEN
    expect(pipelineStack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
        Stages: assert_1.arrayWith({
            Name: 'UpdatePipeline',
            Actions: [
                assert_1.objectLike({
                    Name: 'SelfMutate',
                    Configuration: assert_1.objectLike({
                        ProjectName: { Ref: assert_1.anything() },
                    }),
                }),
            ],
        }),
    });
    expect(pipelineStack).toHaveResourceLike('AWS::CodeBuild::Project', {
        Environment: {
            Image: 'aws/codebuild/standard:4.0',
        },
        Source: {
            BuildSpec: assert_1.encodedJson(assert_1.deepObjectLike({
                phases: {
                    install: {
                        commands: 'npm install -g aws-cdk',
                    },
                    build: {
                        commands: assert_1.arrayWith('cdk -a . deploy PipelineStack --require-approval=never --verbose'),
                    },
                },
            })),
            Type: 'CODEPIPELINE',
        },
    });
});
test('selfmutation stage correctly identifies nested assembly of pipeline stack', () => {
    const pipelineStage = new core_1.Stage(app, 'PipelineStage');
    const nestedPipelineStack = new core_1.Stack(pipelineStage, 'PipelineStack', { env: testutil_1.PIPELINE_ENV });
    new testutil_1.TestGitHubNpmPipeline(nestedPipelineStack, 'Cdk');
    // THEN
    expect(testutil_1.stackTemplate(nestedPipelineStack)).toHaveResourceLike('AWS::CodeBuild::Project', {
        Environment: {
            Image: 'aws/codebuild/standard:4.0',
        },
        Source: {
            BuildSpec: assert_1.encodedJson(assert_1.deepObjectLike({
                phases: {
                    build: {
                        commands: assert_1.arrayWith('cdk -a assembly-PipelineStage deploy PipelineStage-PipelineStack --require-approval=never --verbose'),
                    },
                },
            })),
        },
    });
});
test('overridden stack names are respected', () => {
    // WHEN
    pipeline.addApplicationStage(new OneStackAppWithCustomName(app, 'App1'));
    pipeline.addApplicationStage(new OneStackAppWithCustomName(app, 'App2'));
    // THEN
    expect(pipelineStack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
        Stages: assert_1.arrayWith({
            Name: 'App1',
            Actions: assert_1.arrayWith(assert_1.objectLike({
                Name: 'MyFancyStack.Prepare',
                Configuration: assert_1.objectLike({
                    StackName: 'MyFancyStack',
                }),
            })),
        }, {
            Name: 'App2',
            Actions: assert_1.arrayWith(assert_1.objectLike({
                Name: 'MyFancyStack.Prepare',
                Configuration: assert_1.objectLike({
                    StackName: 'MyFancyStack',
                }),
            })),
        }),
    });
});
test('can control fix/CLI version used in pipeline selfupdate', () => {
    // WHEN
    const stack2 = new core_1.Stack(app, 'Stack2', { env: testutil_1.PIPELINE_ENV });
    new testutil_1.TestGitHubNpmPipeline(stack2, 'Cdk2', {
        pipelineName: 'vpipe',
        cdkCliVersion: '1.2.3',
    });
    // THEN
    expect(stack2).toHaveResourceLike('AWS::CodeBuild::Project', {
        Name: 'vpipe-selfupdate',
        Source: {
            BuildSpec: assert_1.encodedJson(assert_1.deepObjectLike({
                phases: {
                    install: {
                        commands: 'npm install -g aws-cdk@1.2.3',
                    },
                },
            })),
        },
    });
});
test('add another action to an existing stage', () => {
    // WHEN
    pipeline.stage('Source').addAction(new cpa.GitHubSourceAction({
        actionName: 'GitHub2',
        oauthToken: core_1.SecretValue.plainText('oops'),
        output: new cp.Artifact(),
        owner: 'OWNER',
        repo: 'REPO',
    }));
    // THEN
    expect(pipelineStack).toHaveResourceLike('AWS::CodePipeline::Pipeline', {
        Stages: assert_1.arrayWith({
            Name: 'Source',
            Actions: [
                assert_1.objectLike({ Name: 'GitHub' }),
                assert_1.objectLike({ Name: 'GitHub2' }),
            ],
        }),
    });
});
class OneStackApp extends core_1.Stage {
    constructor(scope, id, props) {
        super(scope, id, props);
        new testutil_1.BucketStack(this, 'Stack');
    }
}
class OneStackAppWithCustomName extends core_1.Stage {
    constructor(scope, id, props) {
        super(scope, id, props);
        new testutil_1.BucketStack(this, 'Stack', {
            stackName: 'MyFancyStack',
        });
    }
}
