// @aws-cdk/pipelines/test/testutil.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.stackTemplate = exports.rimraf = exports.BucketStack = exports.TestGitHubAction = exports.TestGitHubNpmPipeline = exports.TestApp = exports.PIPELINE_ENV = void 0;
const fs = require("fs");
const path = require("path");
const codepipeline = require("@aws-cdk/aws-codepipeline");
const codepipeline_actions = require("@aws-cdk/aws-codepipeline-actions");
const s3 = require("@aws-cdk/aws-s3");
const core_1 = require("@aws-cdk/core");
const cdkp = require("../lib");
const construct_internals_1 = require("../lib/private/construct-internals");
exports.PIPELINE_ENV = {
    account: '123pipeline',
    region: 'us-pipeline',
};
class TestApp extends core_1.App {
    constructor(props) {
        super({
            context: {
                '@aws-cdk/core:newStyleStackSynthesis': '1',
            },
            stackTraces: false,
            autoSynth: false,
            runtimeInfo: false,
            treeMetadata: false,
            ...props,
        });
    }
    cleanup() {
        rimraf(construct_internals_1.assemblyBuilderOf(this).outdir);
    }
}
exports.TestApp = TestApp;
class TestGitHubNpmPipeline extends cdkp.CdkPipeline {
    constructor(scope, id, props) {
        var _a, _b;
        const sourceArtifact = (_a = props === null || props === void 0 ? void 0 : props.sourceArtifact) !== null && _a !== void 0 ? _a : new codepipeline.Artifact();
        const cloudAssemblyArtifact = (_b = props === null || props === void 0 ? void 0 : props.cloudAssemblyArtifact) !== null && _b !== void 0 ? _b : new codepipeline.Artifact();
        super(scope, id, {
            sourceAction: new TestGitHubAction(sourceArtifact),
            synthAction: cdkp.SimpleSynthAction.standardNpmSynth({
                sourceArtifact,
                cloudAssemblyArtifact,
            }),
            cloudAssemblyArtifact,
            ...props,
        });
        this.sourceArtifact = sourceArtifact;
        this.cloudAssemblyArtifact = cloudAssemblyArtifact;
    }
}
exports.TestGitHubNpmPipeline = TestGitHubNpmPipeline;
class TestGitHubAction extends codepipeline_actions.GitHubSourceAction {
    constructor(sourceArtifact) {
        super({
            actionName: 'GitHub',
            output: sourceArtifact,
            oauthToken: core_1.SecretValue.plainText('$3kr1t'),
            owner: 'test',
            repo: 'test',
            trigger: codepipeline_actions.GitHubTrigger.POLL,
        });
    }
}
exports.TestGitHubAction = TestGitHubAction;
/**
 * A test stack
 *
 * It contains a single Bucket. Such robust. Much uptime.
 */
class BucketStack extends core_1.Stack {
    constructor(scope, id, props) {
        super(scope, id, props);
        this.bucket = new s3.Bucket(this, 'Bucket');
    }
}
exports.BucketStack = BucketStack;
/**
 * rm -rf reimplementation, don't want to depend on an NPM package for this
 */
function rimraf(fsPath) {
    try {
        const isDir = fs.lstatSync(fsPath).isDirectory();
        if (isDir) {
            for (const file of fs.readdirSync(fsPath)) {
                rimraf(path.join(fsPath, file));
            }
            fs.rmdirSync(fsPath);
        }
        else {
            fs.unlinkSync(fsPath);
        }
    }
    catch (e) {
        // We will survive ENOENT
        if (e.code !== 'ENOENT') {
            throw e;
        }
    }
}
exports.rimraf = rimraf;
/**
 * Because 'expect(stack)' doesn't work correctly for stacks in nested assemblies
 */
function stackTemplate(stack) {
    const stage = core_1.Stage.of(stack);
    if (!stage) {
        throw new Error('stack not in a Stage');
    }
    return stage.synth().getStackArtifact(stack.artifactId);
}
exports.stackTemplate = stackTemplate;
