// @aws-cdk/aws-codebuild/test/test.linux-gpu-build-image.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/core");
const codebuild = require("../lib");
module.exports = {
    'Linux GPU build image': {
        'AWS Deep Learning Container images': {
            'allows passing the account that the repository of the image is hosted in'(test) {
                const stack = new cdk.Stack();
                new codebuild.Project(stack, 'Project', {
                    buildSpec: codebuild.BuildSpec.fromObject({
                        version: '0.2',
                        phases: {
                            build: { commands: ['ls'] },
                        },
                    }),
                    environment: {
                        buildImage: codebuild.LinuxGpuBuildImage.awsDeepLearningContainersImage('my-repo', 'my-tag', '123456789012'),
                    },
                });
                assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CodeBuild::Project', {
                    Environment: {
                        ComputeType: 'BUILD_GENERAL1_LARGE',
                        Image: {
                            'Fn::Join': ['', [
                                    '123456789012.dkr.ecr.',
                                    { Ref: 'AWS::Region' },
                                    '.',
                                    { Ref: 'AWS::URLSuffix' },
                                    '/my-repo:my-tag',
                                ]],
                        },
                    },
                }));
                assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::IAM::Policy', {
                    PolicyDocument: {
                        Statement: assert_1.arrayWith(assert_1.objectLike({
                            Action: [
                                'ecr:BatchCheckLayerAvailability',
                                'ecr:GetDownloadUrlForLayer',
                                'ecr:BatchGetImage',
                            ],
                            Resource: {
                                'Fn::Join': ['', [
                                        'arn:',
                                        { Ref: 'AWS::Partition' },
                                        ':ecr:',
                                        { Ref: 'AWS::Region' },
                                        ':123456789012:repository/my-repo',
                                    ]],
                            },
                        })),
                    },
                }));
                test.done();
            },
        },
    },
};
