// @aws-cdk/aws-cloudfront/test/integ.cloudfront-custom-s3.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const s3 = require("@aws-cdk/aws-s3");
const core_1 = require("@aws-cdk/core");
const cloudfront = require("../lib");
class TestStack extends core_1.Stack {
    constructor(scope, id) {
        super(scope, id);
        const bucket = new s3.Bucket(this, 'Bucket', {
            publicReadAccess: true,
            websiteIndexDocument: 'index.html',
            websiteErrorDocument: '404.html',
        });
        new cloudfront.CloudFrontWebDistribution(this, 'Distribution', {
            viewerProtocolPolicy: cloudfront.ViewerProtocolPolicy.REDIRECT_TO_HTTPS,
            priceClass: cloudfront.PriceClass.PRICE_CLASS_200,
            originConfigs: [
                {
                    behaviors: [{ isDefaultBehavior: true }],
                    customOriginSource: {
                        originProtocolPolicy: cloudfront.OriginProtocolPolicy.HTTP_ONLY,
                        domainName: bucket.bucketWebsiteDomainName,
                    },
                },
            ],
        });
    }
}
const app = new core_1.App();
new TestStack(app, 'cloudfront-custom-s3-integ');
app.synth();
