// @aws-cdk/aws-cloudfront/test/web_distribution.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = require("@aws-cdk/assert");
const certificatemanager = require("@aws-cdk/aws-certificatemanager");
const lambda = require("@aws-cdk/aws-lambda");
const s3 = require("@aws-cdk/aws-s3");
const cdk = require("@aws-cdk/core");
const nodeunit_shim_1 = require("nodeunit-shim");
const lib_1 = require("../lib");
/* eslint-disable quote-props */
nodeunit_shim_1.nodeunitShim({
    'distribution with custom origin adds custom origin'(test) {
        const stack = new cdk.Stack();
        new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
            originConfigs: [
                {
                    originHeaders: {
                        'X-Custom-Header': 'somevalue',
                    },
                    customOriginSource: {
                        domainName: 'myorigin.com',
                    },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                        },
                    ],
                },
            ],
        });
        assert_1.expect(stack).toMatch({
            'Resources': {
                'AnAmazingWebsiteProbablyCFDistribution47E3983B': {
                    'Type': 'AWS::CloudFront::Distribution',
                    'Properties': {
                        'DistributionConfig': {
                            'DefaultCacheBehavior': {
                                'AllowedMethods': [
                                    'GET',
                                    'HEAD',
                                ],
                                'CachedMethods': [
                                    'GET',
                                    'HEAD',
                                ],
                                'ForwardedValues': {
                                    'Cookies': {
                                        'Forward': 'none',
                                    },
                                    'QueryString': false,
                                },
                                'TargetOriginId': 'origin1',
                                'ViewerProtocolPolicy': 'redirect-to-https',
                                'Compress': true,
                            },
                            'DefaultRootObject': 'index.html',
                            'Enabled': true,
                            'HttpVersion': 'http2',
                            'IPV6Enabled': true,
                            'Origins': [
                                {
                                    'CustomOriginConfig': {
                                        'HTTPPort': 80,
                                        'HTTPSPort': 443,
                                        'OriginKeepaliveTimeout': 5,
                                        'OriginProtocolPolicy': 'https-only',
                                        'OriginReadTimeout': 30,
                                        'OriginSSLProtocols': [
                                            'TLSv1.2',
                                        ],
                                    },
                                    'ConnectionAttempts': 3,
                                    'ConnectionTimeout': 10,
                                    'DomainName': 'myorigin.com',
                                    'Id': 'origin1',
                                    'OriginCustomHeaders': [
                                        {
                                            'HeaderName': 'X-Custom-Header',
                                            'HeaderValue': 'somevalue',
                                        },
                                    ],
                                },
                            ],
                            'PriceClass': 'PriceClass_100',
                            'ViewerCertificate': {
                                'CloudFrontDefaultCertificate': true,
                            },
                        },
                    },
                },
            },
        });
        test.done();
    },
    'most basic distribution'(test) {
        const stack = new cdk.Stack();
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
            originConfigs: [
                {
                    s3OriginSource: {
                        s3BucketSource: sourceBucket,
                    },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                        },
                    ],
                },
            ],
        });
        assert_1.expect(stack).toMatch({
            'Resources': {
                'Bucket83908E77': {
                    'Type': 'AWS::S3::Bucket',
                    'DeletionPolicy': 'Retain',
                    'UpdateReplacePolicy': 'Retain',
                },
                'AnAmazingWebsiteProbablyCFDistribution47E3983B': {
                    'Type': 'AWS::CloudFront::Distribution',
                    'Properties': {
                        'DistributionConfig': {
                            'DefaultRootObject': 'index.html',
                            'Origins': [
                                {
                                    'ConnectionAttempts': 3,
                                    'ConnectionTimeout': 10,
                                    'DomainName': {
                                        'Fn::GetAtt': [
                                            'Bucket83908E77',
                                            'RegionalDomainName',
                                        ],
                                    },
                                    'Id': 'origin1',
                                    'S3OriginConfig': {},
                                },
                            ],
                            'ViewerCertificate': {
                                'CloudFrontDefaultCertificate': true,
                            },
                            'PriceClass': 'PriceClass_100',
                            'DefaultCacheBehavior': {
                                'AllowedMethods': [
                                    'GET',
                                    'HEAD',
                                ],
                                'CachedMethods': [
                                    'GET',
                                    'HEAD',
                                ],
                                'TargetOriginId': 'origin1',
                                'ViewerProtocolPolicy': 'redirect-to-https',
                                'ForwardedValues': {
                                    'QueryString': false,
                                    'Cookies': { 'Forward': 'none' },
                                },
                                'Compress': true,
                            },
                            'Enabled': true,
                            'IPV6Enabled': true,
                            'HttpVersion': 'http2',
                        },
                    },
                },
            },
        });
        test.done();
    },
    'distribution with trusted signers on default distribution'(test) {
        const stack = new cdk.Stack();
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
            originConfigs: [
                {
                    s3OriginSource: {
                        s3BucketSource: sourceBucket,
                    },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                            trustedSigners: ['1234'],
                        },
                    ],
                },
            ],
        });
        assert_1.expect(stack).toMatch({
            'Resources': {
                'Bucket83908E77': {
                    'Type': 'AWS::S3::Bucket',
                    'DeletionPolicy': 'Retain',
                    'UpdateReplacePolicy': 'Retain',
                },
                'AnAmazingWebsiteProbablyCFDistribution47E3983B': {
                    'Type': 'AWS::CloudFront::Distribution',
                    'Properties': {
                        'DistributionConfig': {
                            'DefaultRootObject': 'index.html',
                            'Origins': [
                                {
                                    'ConnectionAttempts': 3,
                                    'ConnectionTimeout': 10,
                                    'DomainName': {
                                        'Fn::GetAtt': [
                                            'Bucket83908E77',
                                            'RegionalDomainName',
                                        ],
                                    },
                                    'Id': 'origin1',
                                    'S3OriginConfig': {},
                                },
                            ],
                            'ViewerCertificate': {
                                'CloudFrontDefaultCertificate': true,
                            },
                            'PriceClass': 'PriceClass_100',
                            'DefaultCacheBehavior': {
                                'AllowedMethods': [
                                    'GET',
                                    'HEAD',
                                ],
                                'CachedMethods': [
                                    'GET',
                                    'HEAD',
                                ],
                                'TargetOriginId': 'origin1',
                                'ViewerProtocolPolicy': 'redirect-to-https',
                                'ForwardedValues': {
                                    'QueryString': false,
                                    'Cookies': { 'Forward': 'none' },
                                },
                                'TrustedSigners': [
                                    '1234',
                                ],
                                'Compress': true,
                            },
                            'Enabled': true,
                            'IPV6Enabled': true,
                            'HttpVersion': 'http2',
                        },
                    },
                },
            },
        });
        test.done();
    },
    'distribution with ViewerProtocolPolicy set to a non-default value'(test) {
        const stack = new cdk.Stack();
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
            viewerProtocolPolicy: lib_1.ViewerProtocolPolicy.ALLOW_ALL,
            originConfigs: [
                {
                    s3OriginSource: {
                        s3BucketSource: sourceBucket,
                    },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                        },
                    ],
                },
            ],
        });
        assert_1.expect(stack).toMatch({
            'Resources': {
                'Bucket83908E77': {
                    'Type': 'AWS::S3::Bucket',
                    'DeletionPolicy': 'Retain',
                    'UpdateReplacePolicy': 'Retain',
                },
                'AnAmazingWebsiteProbablyCFDistribution47E3983B': {
                    'Type': 'AWS::CloudFront::Distribution',
                    'Properties': {
                        'DistributionConfig': {
                            'DefaultRootObject': 'index.html',
                            'Origins': [
                                {
                                    'ConnectionAttempts': 3,
                                    'ConnectionTimeout': 10,
                                    'DomainName': {
                                        'Fn::GetAtt': [
                                            'Bucket83908E77',
                                            'RegionalDomainName',
                                        ],
                                    },
                                    'Id': 'origin1',
                                    'S3OriginConfig': {},
                                },
                            ],
                            'ViewerCertificate': {
                                'CloudFrontDefaultCertificate': true,
                            },
                            'PriceClass': 'PriceClass_100',
                            'DefaultCacheBehavior': {
                                'AllowedMethods': [
                                    'GET',
                                    'HEAD',
                                ],
                                'CachedMethods': [
                                    'GET',
                                    'HEAD',
                                ],
                                'TargetOriginId': 'origin1',
                                'ViewerProtocolPolicy': 'allow-all',
                                'ForwardedValues': {
                                    'QueryString': false,
                                    'Cookies': { 'Forward': 'none' },
                                },
                                'Compress': true,
                            },
                            'Enabled': true,
                            'IPV6Enabled': true,
                            'HttpVersion': 'http2',
                        },
                    },
                },
            },
        });
        test.done();
    },
    'distribution with disabled compression'(test) {
        const stack = new cdk.Stack();
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
            originConfigs: [
                {
                    s3OriginSource: {
                        s3BucketSource: sourceBucket,
                    },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                            compress: false,
                        },
                    ],
                },
            ],
        });
        assert_1.expect(stack).toMatch({
            'Resources': {
                'Bucket83908E77': {
                    'Type': 'AWS::S3::Bucket',
                    'DeletionPolicy': 'Retain',
                    'UpdateReplacePolicy': 'Retain',
                },
                'AnAmazingWebsiteProbablyCFDistribution47E3983B': {
                    'Type': 'AWS::CloudFront::Distribution',
                    'Properties': {
                        'DistributionConfig': {
                            'DefaultRootObject': 'index.html',
                            'Origins': [
                                {
                                    'ConnectionAttempts': 3,
                                    'ConnectionTimeout': 10,
                                    'DomainName': {
                                        'Fn::GetAtt': [
                                            'Bucket83908E77',
                                            'RegionalDomainName',
                                        ],
                                    },
                                    'Id': 'origin1',
                                    'S3OriginConfig': {},
                                },
                            ],
                            'ViewerCertificate': {
                                'CloudFrontDefaultCertificate': true,
                            },
                            'PriceClass': 'PriceClass_100',
                            'DefaultCacheBehavior': {
                                'AllowedMethods': [
                                    'GET',
                                    'HEAD',
                                ],
                                'CachedMethods': [
                                    'GET',
                                    'HEAD',
                                ],
                                'TargetOriginId': 'origin1',
                                'ViewerProtocolPolicy': 'redirect-to-https',
                                'ForwardedValues': {
                                    'QueryString': false,
                                    'Cookies': { 'Forward': 'none' },
                                },
                                'Compress': false,
                            },
                            'Enabled': true,
                            'IPV6Enabled': true,
                            'HttpVersion': 'http2',
                        },
                    },
                },
            },
        });
        test.done();
    },
    'distribution with resolvable lambda-association'(test) {
        const stack = new cdk.Stack();
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        const lambdaFunction = new lambda.Function(stack, 'Lambda', {
            code: lambda.Code.inline('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
            originConfigs: [
                {
                    s3OriginSource: {
                        s3BucketSource: sourceBucket,
                    },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                            lambdaFunctionAssociations: [{
                                    eventType: lib_1.LambdaEdgeEventType.ORIGIN_REQUEST,
                                    lambdaFunction: lambdaFunction.currentVersion,
                                    includeBody: true,
                                }],
                        },
                    ],
                },
            ],
        });
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
            'DistributionConfig': {
                'DefaultCacheBehavior': {
                    'LambdaFunctionAssociations': [
                        {
                            'EventType': 'origin-request',
                            'IncludeBody': true,
                            'LambdaFunctionARN': {
                                'Ref': 'LambdaCurrentVersionDF706F6A97fb843e9bd06fcd2bb15eeace80e13e',
                            },
                        },
                    ],
                },
            },
        }));
        test.done();
    },
    'associate a lambda with removable env vars'(test) {
        const app = new cdk.App();
        const stack = new cdk.Stack(app, 'Stack');
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        const lambdaFunction = new lambda.Function(stack, 'Lambda', {
            code: lambda.Code.inline('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        lambdaFunction.addEnvironment('KEY', 'value', { removeInEdge: true });
        new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
            originConfigs: [
                {
                    s3OriginSource: {
                        s3BucketSource: sourceBucket,
                    },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                            lambdaFunctionAssociations: [{
                                    eventType: lib_1.LambdaEdgeEventType.ORIGIN_REQUEST,
                                    lambdaFunction: lambdaFunction.currentVersion,
                                }],
                        },
                    ],
                },
            ],
        });
        assert_1.expect(stack).to(assert_1.haveResource('AWS::Lambda::Function', {
            Environment: assert_1.ABSENT,
        }));
        test.done();
    },
    'throws when associating a lambda with incompatible env vars'(test) {
        const app = new cdk.App();
        const stack = new cdk.Stack(app, 'Stack');
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        const lambdaFunction = new lambda.Function(stack, 'Lambda', {
            code: lambda.Code.inline('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
            environment: {
                KEY: 'value',
            },
        });
        new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
            originConfigs: [
                {
                    s3OriginSource: {
                        s3BucketSource: sourceBucket,
                    },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                            lambdaFunctionAssociations: [{
                                    eventType: lib_1.LambdaEdgeEventType.ORIGIN_REQUEST,
                                    lambdaFunction: lambdaFunction.currentVersion,
                                }],
                        },
                    ],
                },
            ],
        });
        test.throws(() => app.synth(), /KEY/);
        test.done();
    },
    'throws when associating a lambda with includeBody and a response event type'(test) {
        const app = new cdk.App();
        const stack = new cdk.Stack(app, 'Stack');
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        const fnVersion = lambda.Version.fromVersionArn(stack, 'Version', 'arn:aws:lambda:testregion:111111111111:function:myTestFun:v1');
        test.throws(() => {
            new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                originConfigs: [
                    {
                        s3OriginSource: {
                            s3BucketSource: sourceBucket,
                        },
                        behaviors: [
                            {
                                isDefaultBehavior: true,
                                lambdaFunctionAssociations: [{
                                        eventType: lib_1.LambdaEdgeEventType.VIEWER_RESPONSE,
                                        includeBody: true,
                                        lambdaFunction: fnVersion,
                                    }],
                            },
                        ],
                    },
                ],
            });
        }, /'includeBody' can only be true for ORIGIN_REQUEST or VIEWER_REQUEST event types./);
        test.done();
    },
    'distribution has a defaultChild'(test) {
        const stack = new cdk.Stack();
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        const distribution = new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
            originConfigs: [
                {
                    s3OriginSource: {
                        s3BucketSource: sourceBucket,
                    },
                    behaviors: [{ isDefaultBehavior: true }],
                },
            ],
        });
        test.ok(distribution.node.defaultChild instanceof lib_1.CfnDistribution);
        test.done();
    },
    'allows multiple aliasConfiguration CloudFrontWebDistribution per stack'(test) {
        const stack = new cdk.Stack();
        const s3BucketSource = new s3.Bucket(stack, 'Bucket');
        const originConfigs = [{
                s3OriginSource: { s3BucketSource },
                behaviors: [{ isDefaultBehavior: true }],
            }];
        new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
            originConfigs,
            aliasConfiguration: { acmCertRef: 'acm_ref', names: ['www.example.com'] },
        });
        new lib_1.CloudFrontWebDistribution(stack, 'AnotherAmazingWebsiteProbably', {
            originConfigs,
            aliasConfiguration: { acmCertRef: 'another_acm_ref', names: ['ftp.example.com'] },
        });
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
            'DistributionConfig': {
                'Aliases': ['www.example.com'],
                'ViewerCertificate': {
                    'AcmCertificateArn': 'acm_ref',
                    'SslSupportMethod': 'sni-only',
                },
            },
        }));
        assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
            'DistributionConfig': {
                'Aliases': ['ftp.example.com'],
                'ViewerCertificate': {
                    'AcmCertificateArn': 'another_acm_ref',
                    'SslSupportMethod': 'sni-only',
                },
            },
        }));
        test.done();
    },
    'viewerCertificate': {
        'acmCertificate': {
            'base usage'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                const certificate = new certificatemanager.Certificate(stack, 'cert', {
                    domainName: 'example.com',
                });
                new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                    originConfigs: [{
                            s3OriginSource: { s3BucketSource: sourceBucket },
                            behaviors: [{ isDefaultBehavior: true }],
                        }],
                    viewerCertificate: lib_1.ViewerCertificate.fromAcmCertificate(certificate),
                });
                assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
                    'DistributionConfig': {
                        'Aliases': [],
                        'ViewerCertificate': {
                            'AcmCertificateArn': {
                                'Ref': 'cert56CA94EB',
                            },
                            'SslSupportMethod': 'sni-only',
                        },
                    },
                }));
                test.done();
            },
            'imported certificate fromCertificateArn'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                const certificate = certificatemanager.Certificate.fromCertificateArn(stack, 'cert', 'arn:aws:acm:us-east-1:1111111:certificate/11-3336f1-44483d-adc7-9cd375c5169d');
                new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                    originConfigs: [{
                            s3OriginSource: { s3BucketSource: sourceBucket },
                            behaviors: [{ isDefaultBehavior: true }],
                        }],
                    viewerCertificate: lib_1.ViewerCertificate.fromAcmCertificate(certificate),
                });
                assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
                    'DistributionConfig': {
                        'Aliases': [],
                        'ViewerCertificate': {
                            'AcmCertificateArn': 'arn:aws:acm:us-east-1:1111111:certificate/11-3336f1-44483d-adc7-9cd375c5169d',
                            'SslSupportMethod': 'sni-only',
                        },
                    },
                }));
                test.done();
            },
            'advanced usage'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                const certificate = new certificatemanager.Certificate(stack, 'cert', {
                    domainName: 'example.com',
                });
                new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                    originConfigs: [{
                            s3OriginSource: { s3BucketSource: sourceBucket },
                            behaviors: [{ isDefaultBehavior: true }],
                        }],
                    viewerCertificate: lib_1.ViewerCertificate.fromAcmCertificate(certificate, {
                        securityPolicy: lib_1.SecurityPolicyProtocol.SSL_V3,
                        sslMethod: lib_1.SSLMethod.VIP,
                        aliases: ['example.com', 'www.example.com'],
                    }),
                });
                assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
                    'DistributionConfig': {
                        'Aliases': ['example.com', 'www.example.com'],
                        'ViewerCertificate': {
                            'AcmCertificateArn': {
                                'Ref': 'cert56CA94EB',
                            },
                            'MinimumProtocolVersion': 'SSLv3',
                            'SslSupportMethod': 'vip',
                        },
                    },
                }));
                test.done();
            },
        },
        'iamCertificate': {
            'base usage'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                    originConfigs: [{
                            s3OriginSource: { s3BucketSource: sourceBucket },
                            behaviors: [{ isDefaultBehavior: true }],
                        }],
                    viewerCertificate: lib_1.ViewerCertificate.fromIamCertificate('test'),
                });
                assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
                    'DistributionConfig': {
                        'Aliases': [],
                        'ViewerCertificate': {
                            'IamCertificateId': 'test',
                            'SslSupportMethod': 'sni-only',
                        },
                    },
                }));
                test.done();
            },
            'advanced usage'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                    originConfigs: [{
                            s3OriginSource: { s3BucketSource: sourceBucket },
                            behaviors: [{ isDefaultBehavior: true }],
                        }],
                    viewerCertificate: lib_1.ViewerCertificate.fromIamCertificate('test', {
                        securityPolicy: lib_1.SecurityPolicyProtocol.TLS_V1,
                        sslMethod: lib_1.SSLMethod.VIP,
                        aliases: ['example.com'],
                    }),
                });
                assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
                    'DistributionConfig': {
                        'Aliases': ['example.com'],
                        'ViewerCertificate': {
                            'IamCertificateId': 'test',
                            'MinimumProtocolVersion': 'TLSv1',
                            'SslSupportMethod': 'vip',
                        },
                    },
                }));
                test.done();
            },
        },
        'cloudFrontDefaultCertificate': {
            'base usage'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                    originConfigs: [{
                            s3OriginSource: { s3BucketSource: sourceBucket },
                            behaviors: [{ isDefaultBehavior: true }],
                        }],
                    viewerCertificate: lib_1.ViewerCertificate.fromCloudFrontDefaultCertificate(),
                });
                assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
                    'DistributionConfig': {
                        'Aliases': [],
                        'ViewerCertificate': {
                            'CloudFrontDefaultCertificate': true,
                        },
                    },
                }));
                test.done();
            },
            'aliases are set'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                    originConfigs: [{
                            s3OriginSource: { s3BucketSource: sourceBucket },
                            behaviors: [{ isDefaultBehavior: true }],
                        }],
                    viewerCertificate: lib_1.ViewerCertificate.fromCloudFrontDefaultCertificate('example.com', 'www.example.com'),
                });
                assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
                    'DistributionConfig': {
                        'Aliases': ['example.com', 'www.example.com'],
                        'ViewerCertificate': {
                            'CloudFrontDefaultCertificate': true,
                        },
                    },
                }));
                test.done();
            },
        },
        'errors': {
            'throws if both deprecated aliasConfiguration and viewerCertificate'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                test.throws(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                        originConfigs: [{
                                s3OriginSource: { s3BucketSource: sourceBucket },
                                behaviors: [{ isDefaultBehavior: true }],
                            }],
                        aliasConfiguration: { acmCertRef: 'test', names: ['ftp.example.com'] },
                        viewerCertificate: lib_1.ViewerCertificate.fromCloudFrontDefaultCertificate('example.com', 'www.example.com'),
                    });
                }, /You cannot set both aliasConfiguration and viewerCertificate properties/);
                test.done();
            },
            'throws if invalid security policy for SSL method'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                test.throws(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                        originConfigs: [{
                                s3OriginSource: { s3BucketSource: sourceBucket },
                                behaviors: [{ isDefaultBehavior: true }],
                            }],
                        viewerCertificate: lib_1.ViewerCertificate.fromIamCertificate('test', {
                            securityPolicy: lib_1.SecurityPolicyProtocol.TLS_V1_1_2016,
                            sslMethod: lib_1.SSLMethod.VIP,
                        }),
                    });
                }, /TLSv1.1_2016 is not compabtible with sslMethod vip./);
                test.done();
            },
            // FIXME https://github.com/aws/aws-cdk/issues/4724
            'does not throw if acmCertificate explicitly not in us-east-1'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                const certificate = certificatemanager.Certificate.fromCertificateArn(stack, 'cert', 'arn:aws:acm:eu-west-3:1111111:certificate/11-3336f1-44483d-adc7-9cd375c5169d');
                new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                    originConfigs: [{
                            s3OriginSource: { s3BucketSource: sourceBucket },
                            behaviors: [{ isDefaultBehavior: true }],
                        }],
                    viewerCertificate: lib_1.ViewerCertificate.fromAcmCertificate(certificate),
                });
                assert_1.expect(stack).to(assert_1.haveResourceLike('AWS::CloudFront::Distribution', {
                    'DistributionConfig': {
                        'Aliases': [],
                        'ViewerCertificate': {
                            'AcmCertificateArn': 'arn:aws:acm:eu-west-3:1111111:certificate/11-3336f1-44483d-adc7-9cd375c5169d',
                            'SslSupportMethod': 'sni-only',
                        },
                    },
                }));
                test.done();
            },
        },
    },
    'edgelambda.amazonaws.com is added to the trust policy of lambda'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        const fn = new lambda.Function(stack, 'Lambda', {
            code: lambda.Code.fromInline('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NODEJS_10_X,
        });
        const lambdaVersion = new lambda.Version(stack, 'LambdaVersion', { lambda: fn });
        // WHEN
        new lib_1.CloudFrontWebDistribution(stack, 'MyDistribution', {
            originConfigs: [
                {
                    s3OriginSource: { s3BucketSource: sourceBucket },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                            lambdaFunctionAssociations: [
                                {
                                    eventType: lib_1.LambdaEdgeEventType.ORIGIN_REQUEST,
                                    lambdaFunction: lambdaVersion,
                                },
                            ],
                        },
                    ],
                },
            ],
        });
        // THEN
        assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Role', {
            AssumeRolePolicyDocument: {
                'Statement': [
                    {
                        'Action': 'sts:AssumeRole',
                        'Effect': 'Allow',
                        'Principal': {
                            'Service': 'lambda.amazonaws.com',
                        },
                    },
                    {
                        'Action': 'sts:AssumeRole',
                        'Effect': 'Allow',
                        'Principal': {
                            'Service': 'edgelambda.amazonaws.com',
                        },
                    },
                ],
                'Version': '2012-10-17',
            },
        }));
        test.done();
    },
    'edgelambda.amazonaws.com is not added to lambda role for imported functions'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const sourceBucket = new s3.Bucket(stack, 'Bucket');
        const lambdaVersion = lambda.Version.fromVersionArn(stack, 'Version', 'arn:my-version');
        // WHEN
        new lib_1.CloudFrontWebDistribution(stack, 'MyDistribution', {
            originConfigs: [
                {
                    s3OriginSource: { s3BucketSource: sourceBucket },
                    behaviors: [
                        {
                            isDefaultBehavior: true,
                            lambdaFunctionAssociations: [
                                {
                                    eventType: lib_1.LambdaEdgeEventType.ORIGIN_REQUEST,
                                    lambdaFunction: lambdaVersion,
                                },
                            ],
                        },
                    ],
                },
            ],
        });
        assert_1.expect(stack).notTo(assert_1.haveResourceLike('AWS::IAM::Role'));
        test.done();
    },
    'geo restriction': {
        'success': {
            'whitelist'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                    originConfigs: [{
                            s3OriginSource: { s3BucketSource: sourceBucket },
                            behaviors: [{ isDefaultBehavior: true }],
                        }],
                    geoRestriction: lib_1.GeoRestriction.whitelist('US', 'UK'),
                });
                assert_1.expect(stack).toMatch({
                    'Resources': {
                        'Bucket83908E77': {
                            'Type': 'AWS::S3::Bucket',
                            'DeletionPolicy': 'Retain',
                            'UpdateReplacePolicy': 'Retain',
                        },
                        'AnAmazingWebsiteProbablyCFDistribution47E3983B': {
                            'Type': 'AWS::CloudFront::Distribution',
                            'Properties': {
                                'DistributionConfig': {
                                    'DefaultRootObject': 'index.html',
                                    'Origins': [
                                        {
                                            'ConnectionAttempts': 3,
                                            'ConnectionTimeout': 10,
                                            'DomainName': {
                                                'Fn::GetAtt': [
                                                    'Bucket83908E77',
                                                    'RegionalDomainName',
                                                ],
                                            },
                                            'Id': 'origin1',
                                            'S3OriginConfig': {},
                                        },
                                    ],
                                    'ViewerCertificate': {
                                        'CloudFrontDefaultCertificate': true,
                                    },
                                    'PriceClass': 'PriceClass_100',
                                    'DefaultCacheBehavior': {
                                        'AllowedMethods': [
                                            'GET',
                                            'HEAD',
                                        ],
                                        'CachedMethods': [
                                            'GET',
                                            'HEAD',
                                        ],
                                        'TargetOriginId': 'origin1',
                                        'ViewerProtocolPolicy': 'redirect-to-https',
                                        'ForwardedValues': {
                                            'QueryString': false,
                                            'Cookies': { 'Forward': 'none' },
                                        },
                                        'Compress': true,
                                    },
                                    'Enabled': true,
                                    'IPV6Enabled': true,
                                    'HttpVersion': 'http2',
                                    'Restrictions': {
                                        'GeoRestriction': {
                                            'Locations': ['US', 'UK'],
                                            'RestrictionType': 'whitelist',
                                        },
                                    },
                                },
                            },
                        },
                    },
                });
                test.done();
            },
            'blacklist'(test) {
                const stack = new cdk.Stack();
                const sourceBucket = new s3.Bucket(stack, 'Bucket');
                new lib_1.CloudFrontWebDistribution(stack, 'AnAmazingWebsiteProbably', {
                    originConfigs: [{
                            s3OriginSource: { s3BucketSource: sourceBucket },
                            behaviors: [{ isDefaultBehavior: true }],
                        }],
                    geoRestriction: lib_1.GeoRestriction.blacklist('US'),
                });
                assert_1.expect(stack).toMatch({
                    'Resources': {
                        'Bucket83908E77': {
                            'Type': 'AWS::S3::Bucket',
                            'DeletionPolicy': 'Retain',
                            'UpdateReplacePolicy': 'Retain',
                        },
                        'AnAmazingWebsiteProbablyCFDistribution47E3983B': {
                            'Type': 'AWS::CloudFront::Distribution',
                            'Properties': {
                                'DistributionConfig': {
                                    'DefaultRootObject': 'index.html',
                                    'Origins': [
                                        {
                                            'ConnectionAttempts': 3,
                                            'ConnectionTimeout': 10,
                                            'DomainName': {
                                                'Fn::GetAtt': [
                                                    'Bucket83908E77',
                                                    'RegionalDomainName',
                                                ],
                                            },
                                            'Id': 'origin1',
                                            'S3OriginConfig': {},
                                        },
                                    ],
                                    'ViewerCertificate': {
                                        'CloudFrontDefaultCertificate': true,
                                    },
                                    'PriceClass': 'PriceClass_100',
                                    'DefaultCacheBehavior': {
                                        'AllowedMethods': [
                                            'GET',
                                            'HEAD',
                                        ],
                                        'CachedMethods': [
                                            'GET',
                                            'HEAD',
                                        ],
                                        'TargetOriginId': 'origin1',
                                        'ViewerProtocolPolicy': 'redirect-to-https',
                                        'ForwardedValues': {
                                            'QueryString': false,
                                            'Cookies': { 'Forward': 'none' },
                                        },
                                        'Compress': true,
                                    },
                                    'Enabled': true,
                                    'IPV6Enabled': true,
                                    'HttpVersion': 'http2',
                                    'Restrictions': {
                                        'GeoRestriction': {
                                            'Locations': ['US'],
                                            'RestrictionType': 'blacklist',
                                        },
                                    },
                                },
                            },
                        },
                    },
                });
                test.done();
            },
        },
        'error': {
            'throws if locations is empty array'(test) {
                test.throws(() => {
                    lib_1.GeoRestriction.whitelist();
                }, /Should provide at least 1 location/);
                test.throws(() => {
                    lib_1.GeoRestriction.blacklist();
                }, /Should provide at least 1 location/);
                test.done();
            },
            'throws if locations format is wrong'(test) {
                test.throws(() => {
                    lib_1.GeoRestriction.whitelist('us');
                }, /Invalid location format for location: us, location should be two-letter and uppercase country ISO 3166-1-alpha-2 code/);
                test.throws(() => {
                    lib_1.GeoRestriction.blacklist('us');
                }, /Invalid location format for location: us, location should be two-letter and uppercase country ISO 3166-1-alpha-2 code/);
                test.done();
            },
        },
    },
    'Connection behaviors between CloudFront and your origin': {
        'success': {
            'connectionAttempts = 1'(test) {
                const stack = new cdk.Stack();
                test.doesNotThrow(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionAttempts: 1,
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionAttempts: You can specify 1, 2, or 3 as the number of attempts./);
                test.done();
            },
            '3 = connectionAttempts'(test) {
                const stack = new cdk.Stack();
                test.doesNotThrow(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionAttempts: 3,
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionAttempts: You can specify 1, 2, or 3 as the number of attempts./);
                test.done();
            },
            'connectionTimeout = 1'(test) {
                const stack = new cdk.Stack();
                test.doesNotThrow(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionTimeout: cdk.Duration.seconds(1),
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionTimeout: You can specify a number of seconds between 1 and 10 (inclusive)./);
                test.done();
            },
            '10 = connectionTimeout'(test) {
                const stack = new cdk.Stack();
                test.doesNotThrow(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionTimeout: cdk.Duration.seconds(10),
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionTimeout: You can specify a number of seconds between 1 and 10 (inclusive)./);
                test.done();
            },
        },
        'errors': {
            'connectionAttempts = 1.1'(test) {
                const stack = new cdk.Stack();
                test.throws(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionAttempts: 1.1,
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionAttempts: You can specify 1, 2, or 3 as the number of attempts./);
                test.done();
            },
            'connectionAttempts = -1'(test) {
                const stack = new cdk.Stack();
                test.throws(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionAttempts: -1,
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionAttempts: You can specify 1, 2, or 3 as the number of attempts./);
                test.done();
            },
            'connectionAttempts < 1'(test) {
                const stack = new cdk.Stack();
                test.throws(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionAttempts: 0,
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionAttempts: You can specify 1, 2, or 3 as the number of attempts./);
                test.done();
            },
            '3 < connectionAttempts'(test) {
                const stack = new cdk.Stack();
                test.throws(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionAttempts: 4,
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionAttempts: You can specify 1, 2, or 3 as the number of attempts./);
                test.done();
            },
            'connectionTimeout = 1.1'(test) {
                const stack = new cdk.Stack();
                test.throws(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionTimeout: cdk.Duration.seconds(1.1),
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionTimeout: You can specify a number of seconds between 1 and 10 \(inclusive\)./);
                test.done();
            },
            'connectionTimeout < 1'(test) {
                const stack = new cdk.Stack();
                test.throws(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionTimeout: cdk.Duration.seconds(0),
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionTimeout: You can specify a number of seconds between 1 and 10 \(inclusive\)./);
                test.done();
            },
            '10 < connectionTimeout'(test) {
                const stack = new cdk.Stack();
                test.throws(() => {
                    new lib_1.CloudFrontWebDistribution(stack, 'Distribution', {
                        originConfigs: [{
                                behaviors: [{ isDefaultBehavior: true }],
                                connectionTimeout: cdk.Duration.seconds(11),
                                customOriginSource: { domainName: 'myorigin.com' },
                            }],
                    });
                }, /connectionTimeout: You can specify a number of seconds between 1 and 10 \(inclusive\)./);
                test.done();
            },
        },
    },
    'existing distributions can be imported'(test) {
        const stack = new cdk.Stack();
        const dist = lib_1.CloudFrontWebDistribution.fromDistributionAttributes(stack, 'ImportedDist', {
            domainName: 'd111111abcdef8.cloudfront.net',
            distributionId: '012345ABCDEF',
        });
        test.equals(dist.distributionDomainName, 'd111111abcdef8.cloudfront.net');
        test.equals(dist.distributionId, '012345ABCDEF');
        test.done();
    },
});
