// @aws-cdk/aws-events-targets/test/aws-api/integ.aws-api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/// !cdk-integ pragma:ignore-assets
const events = require("@aws-cdk/aws-events");
const cdk = require("@aws-cdk/core");
const targets = require("../../lib");
const app = new cdk.App();
class AwsApi extends cdk.Stack {
    constructor(scope, id, props) {
        super(scope, id, props);
        // Force new deployment of 'cool-service' and stop 'dev-instance' at midnight everyday
        const scheduleRule = new events.Rule(this, 'ScheduleRule', {
            schedule: events.Schedule.cron({
                hour: '0',
                minute: '0',
            }),
        });
        scheduleRule.addTarget(new targets.AwsApi({
            service: 'ECS',
            action: 'updateService',
            parameters: {
                service: 'cool-service',
                forceNewDeployment: true,
            },
        }));
        scheduleRule.addTarget(new targets.AwsApi({
            service: 'RDS',
            action: 'stopDBInstance',
            parameters: {
                DBInstanceIdentifier: 'dev-instance',
            },
        }));
        // Create snapshots when a DB instance restarts
        const patternRule = new events.Rule(this, 'PatternRule', {
            eventPattern: {
                detailType: ['RDS DB Instance Event'],
                detail: {
                    Message: ['DB instance restarted'],
                },
            },
        });
        patternRule.addTarget(new targets.AwsApi({
            service: 'RDS',
            action: 'createDBSnapshot',
            parameters: {
                DBInstanceIdentifier: events.EventField.fromPath('$.detail.SourceArn'),
            },
        }));
    }
}
new AwsApi(app, 'aws-cdk-aws-api-target-integ');
app.synth();
