// @aws-cdk/aws-cognito/test/user-pool-idps/facebook.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../../lib");
describe('UserPoolIdentityProvider', () => {
    describe('facebook', () => {
        test('defaults', () => {
            // GIVEN
            const stack = new core_1.Stack();
            const pool = new lib_1.UserPool(stack, 'userpool');
            // WHEN
            new lib_1.UserPoolIdentityProviderFacebook(stack, 'userpoolidp', {
                userPool: pool,
                clientId: 'fb-client-id',
                clientSecret: 'fb-client-secret',
            });
            expect(stack).toHaveResource('AWS::Cognito::UserPoolIdentityProvider', {
                ProviderName: 'Facebook',
                ProviderType: 'Facebook',
                ProviderDetails: {
                    client_id: 'fb-client-id',
                    client_secret: 'fb-client-secret',
                    authorize_scopes: 'public_profile',
                },
            });
        });
        test('scopes & api_version', () => {
            // GIVEN
            const stack = new core_1.Stack();
            const pool = new lib_1.UserPool(stack, 'userpool');
            // WHEN
            new lib_1.UserPoolIdentityProviderFacebook(stack, 'userpoolidp', {
                userPool: pool,
                clientId: 'fb-client-id',
                clientSecret: 'fb-client-secret',
                scopes: ['scope1', 'scope2'],
                apiVersion: 'version1',
            });
            expect(stack).toHaveResource('AWS::Cognito::UserPoolIdentityProvider', {
                ProviderName: 'Facebook',
                ProviderType: 'Facebook',
                ProviderDetails: {
                    client_id: 'fb-client-id',
                    client_secret: 'fb-client-secret',
                    authorize_scopes: 'scope1,scope2',
                    api_version: 'version1',
                },
            });
        });
        test('registered with user pool', () => {
            // GIVEN
            const stack = new core_1.Stack();
            const pool = new lib_1.UserPool(stack, 'userpool');
            // WHEN
            const provider = new lib_1.UserPoolIdentityProviderFacebook(stack, 'userpoolidp', {
                userPool: pool,
                clientId: 'fb-client-id',
                clientSecret: 'fb-client-secret',
            });
            // THEN
            expect(pool.identityProviders).toContain(provider);
        });
        test('attribute mapping', () => {
            // GIVEN
            const stack = new core_1.Stack();
            const pool = new lib_1.UserPool(stack, 'userpool');
            // WHEN
            new lib_1.UserPoolIdentityProviderFacebook(stack, 'userpoolidp', {
                userPool: pool,
                clientId: 'fb-client-id',
                clientSecret: 'fb-client-secret',
                attributeMapping: {
                    givenName: lib_1.ProviderAttribute.FACEBOOK_NAME,
                    address: lib_1.ProviderAttribute.other('fb-address'),
                    custom: {
                        customAttr1: lib_1.ProviderAttribute.FACEBOOK_EMAIL,
                        customAttr2: lib_1.ProviderAttribute.other('fb-custom-attr'),
                    },
                },
            });
            // THEN
            expect(stack).toHaveResource('AWS::Cognito::UserPoolIdentityProvider', {
                AttributeMapping: {
                    given_name: 'name',
                    address: 'fb-address',
                    customAttr1: 'email',
                    customAttr2: 'fb-custom-attr',
                },
            });
        });
    });
});
