// @aws-cdk/aws-cloudfront/test/private/cache-behavior.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("@aws-cdk/assert/jest");
const lambda = require("@aws-cdk/aws-lambda");
const core_1 = require("@aws-cdk/core");
const lib_1 = require("../../lib");
const cache_behavior_1 = require("../../lib/private/cache-behavior");
let app;
let stack;
beforeEach(() => {
    app = new core_1.App();
    stack = new core_1.Stack(app, 'Stack', {
        env: { account: '1234', region: 'testregion' },
    });
});
test('renders the minimum template with an origin and path specified', () => {
    const behavior = new cache_behavior_1.CacheBehavior('origin_id', {
        pathPattern: '*',
    });
    expect(behavior._renderBehavior()).toEqual({
        targetOriginId: 'origin_id',
        pathPattern: '*',
        forwardedValues: { queryString: false },
        viewerProtocolPolicy: 'allow-all',
    });
});
test('renders with all properties specified', () => {
    const fnVersion = lambda.Version.fromVersionArn(stack, 'Version', 'arn:aws:lambda:testregion:111111111111:function:myTestFun:v1');
    const behavior = new cache_behavior_1.CacheBehavior('origin_id', {
        pathPattern: '*',
        allowedMethods: lib_1.AllowedMethods.ALLOW_ALL,
        cachedMethods: lib_1.CachedMethods.CACHE_GET_HEAD_OPTIONS,
        compress: true,
        forwardQueryString: true,
        forwardQueryStringCacheKeys: ['user_id', 'auth'],
        smoothStreaming: true,
        viewerProtocolPolicy: lib_1.ViewerProtocolPolicy.HTTPS_ONLY,
        edgeLambdas: [{
                eventType: lib_1.LambdaEdgeEventType.ORIGIN_REQUEST,
                includeBody: true,
                functionVersion: fnVersion,
            }],
    });
    expect(behavior._renderBehavior()).toEqual({
        targetOriginId: 'origin_id',
        pathPattern: '*',
        allowedMethods: ['GET', 'HEAD', 'OPTIONS', 'PUT', 'PATCH', 'POST', 'DELETE'],
        cachedMethods: ['GET', 'HEAD', 'OPTIONS'],
        compress: true,
        forwardedValues: {
            queryString: true,
            queryStringCacheKeys: ['user_id', 'auth'],
        },
        smoothStreaming: true,
        viewerProtocolPolicy: 'https-only',
        lambdaFunctionAssociations: [{
                lambdaFunctionArn: 'arn:aws:lambda:testregion:111111111111:function:myTestFun:v1',
                eventType: 'origin-request',
                includeBody: true,
            }],
    });
});
test('throws if edgeLambda includeBody is set for wrong event type', () => {
    const fnVersion = lambda.Version.fromVersionArn(stack, 'Version', 'arn:aws:lambda:testregion:111111111111:function:myTestFun:v1');
    expect(() => new cache_behavior_1.CacheBehavior('origin_id', {
        pathPattern: '*',
        edgeLambdas: [{
                eventType: lib_1.LambdaEdgeEventType.ORIGIN_RESPONSE,
                includeBody: true,
                functionVersion: fnVersion,
            }],
    })).toThrow(/'includeBody' can only be true for ORIGIN_REQUEST or VIEWER_REQUEST event types./);
});
