// aws-cdk/test/integ/cli/corking.js
"use strict";
/**
 * Routines for corking stdout and stderr
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.flushCorkedOutput = exports.uncorkShellOutput = exports.writeOutput = exports.corkShellOutput = void 0;
let _corkShellOutput = false;
const _corked = {
    stdout: new Array(),
    stderr: new Array(),
};
function cleanStreams() {
    _corked.stdout.splice(0, _corked.stdout.length);
    _corked.stderr.splice(0, _corked.stderr.length);
}
function corkShellOutput() {
    _corkShellOutput = true;
    cleanStreams();
}
exports.corkShellOutput = corkShellOutput;
function writeOutput(stream, content) {
    if (_corkShellOutput) {
        _corked[stream].push(content);
    }
    else {
        process[stream].write(content);
    }
}
exports.writeOutput = writeOutput;
async function writeAndFlush(stream, content) {
    const flushed = process[stream].write(content);
    if (!flushed) {
        return new Promise(ok => process[stream].once('drain', ok));
    }
}
function uncorkShellOutput() {
    _corkShellOutput = false;
}
exports.uncorkShellOutput = uncorkShellOutput;
async function flushCorkedOutput() {
    await writeAndFlush('stdout', Buffer.concat(_corked.stdout));
    await writeAndFlush('stderr', Buffer.concat(_corked.stderr));
    cleanStreams();
}
exports.flushCorkedOutput = flushCorkedOutput;
